File: backend/src/main/kotlin/com/washop/models/Order.kt
package com.washop.models

import kotlinx.serialization.Serializable
import org.jetbrains.exposed.sql.Table
import org.jetbrains.exposed.sql.javatime.timestamp
import java.time.Instant

object Orders : Table("orders") {
    val id = varchar("id", 50)
    val shopId = varchar("shop_id", 50).references(Shops.id)
    val customerPhone = varchar("customer_phone", 15)
    val customerName = varchar("customer_name", 255).nullable()
    val deliveryAddress = text("delivery_address").nullable()
    val subtotal = decimal("subtotal", 10, 2).default(java.math.BigDecimal.ZERO)
    val deliveryCharge = decimal("delivery_charge", 10, 2).default(java.math.BigDecimal.ZERO)
    val discount = decimal("discount", 10, 2).default(java.math.BigDecimal.ZERO)
    val total = decimal("total", 10, 2).default(java.math.BigDecimal.ZERO)
    val status = varchar("status", 20).default("PENDING")
    val paymentStatus = varchar("payment_status", 20).default("UNPAID")
    val paymentMethod = varchar("payment_method", 20).nullable()
    val notes = text("notes").nullable()
    val createdAt = timestamp("created_at").default(Instant.now())
    val updatedAt = timestamp("updated_at").default(Instant.now())
    val acceptedAt = timestamp("accepted_at").nullable()
    val deliveredAt = timestamp("delivered_at").nullable()

    override val primaryKey = PrimaryKey(id)
}

object OrderItems : Table("order_items") {
    val id = integer("id").autoIncrement()
    val orderId = varchar("order_id", 50).references(Orders.id)
    val productId = varchar("product_id", 50).references(Products.id).nullable()
    val productName = varchar("product_name", 255)
    val quantity = integer("quantity")
    val unitPrice = decimal("unit_price", 10, 2)
    val totalPrice = decimal("total_price", 10, 2)
    val createdAt = timestamp("created_at").default(Instant.now())

    override val primaryKey = PrimaryKey(id)
}

enum class OrderStatus {
    PENDING,
    ACCEPTED,
    REJECTED,
    PREPARING,
    OUT_FOR_DELIVERY,
    DELIVERED,
    CANCELLED
}

enum class PaymentStatus {
    UNPAID,
    PAID,
    REFUNDED,
    COD
}

@Serializable
data class Order(
    val id: String,
    val shopId: String,
    val customerPhone: String,
    val customerName: String? = null,
    val deliveryAddress: String? = null,
    val subtotal: Double,
    val deliveryCharge: Double = 0.0,
    val discount: Double = 0.0,
    val total: Double,
    val status: String = "PENDING",
    val paymentStatus: String = "UNPAID",
    val paymentMethod: String? = null,
    val notes: String? = null,
    val items: List<OrderItemResponse> = emptyList(),
    val createdAt: String? = null
)

@Serializable
data class OrderItemResponse(
    val id: Int,
    val productId: String?,
    val productName: String,
    val quantity: Int,
    val unitPrice: Double,
    val totalPrice: Double
)

@Serializable
data class CreateOrderRequest(
    val shopId: String,
    val customerPhone: String,
    val customerName: String? = null,
    val deliveryAddress: String? = null,
    val items: List<OrderItemRequest>,
    val notes: String? = null,
    val paymentMethod: String? = null
)

@Serializable
data class OrderItemRequest(
    val productId: String,
    val quantity: Int
)

@Serializable
data class UpdateOrderStatusRequest(
    val status: String
)

@Serializable
data class OrderListResponse(
    val orders: List<Order>,
    val total: Int,
    val page: Int,
    val pageSize: Int
)

File: flutter_app/lib/models/order.dart
class Order {
  final String id;
  final String shopId;
  final String customerPhone;
  final String? customerName;
  final String? deliveryAddress;
  final double subtotal;
  final double deliveryCharge;
  final double discount;
  final double total;
  final String status;
  final String paymentStatus;
  final String? paymentMethod;
  final String? notes;
  final List<OrderItem> items;
  final String? createdAt;

  Order({
    required this.id,
    required this.shopId,
    required this.customerPhone,
    this.customerName,
    this.deliveryAddress,
    required this.subtotal,
    this.deliveryCharge = 0,
    this.discount = 0,
    required this.total,
    this.status = 'PENDING',
    this.paymentStatus = 'UNPAID',
    this.paymentMethod,
    this.notes,
    this.items = const [],
    this.createdAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] as String,
      shopId: json['shopId'] as String,
      customerPhone: json['customerPhone'] as String,
      customerName: json['customerName'] as String?,
      deliveryAddress: json['deliveryAddress'] as String?,
      subtotal: (json['subtotal'] as num).toDouble(),
      deliveryCharge: (json['deliveryCharge'] as num?)?.toDouble() ?? 0,
      discount: (json['discount'] as num?)?.toDouble() ?? 0,
      total: (json['total'] as num).toDouble(),
      status: json['status'] as String? ?? 'PENDING',
      paymentStatus: json['paymentStatus'] as String? ?? 'UNPAID',
      paymentMethod: json['paymentMethod'] as String?,
      notes: json['notes'] as String?,
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => OrderItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: json['createdAt'] as String?,
    );
  }

  String get statusDisplay {
    switch (status) {
      case 'PENDING':
        return 'Pending';
      case 'ACCEPTED':
        return 'Accepted';
      case 'REJECTED':
        return 'Rejected';
      case 'PREPARING':
        return 'Preparing';
      case 'OUT_FOR_DELIVERY':
        return 'Out for Delivery';
      case 'DELIVERED':
        return 'Delivered';
      case 'CANCELLED':
        return 'Cancelled';
      default:
        return status;
    }
  }

  String get paymentStatusDisplay {
    switch (paymentStatus) {
      case 'UNPAID':
        return 'Unpaid';
      case 'PAID':
        return 'Paid';
      case 'REFUNDED':
        return 'Refunded';
      case 'COD':
        return 'Cash on Delivery';
      default:
        return paymentStatus;
    }
  }

  bool get isPending => status == 'PENDING';
  bool get isAccepted => status == 'ACCEPTED';
  bool get isDelivered => status == 'DELIVERED';
  bool get isCancelled => status == 'CANCELLED' || status == 'REJECTED';
}

class OrderItem {
  final int id;
  final String? productId;
  final String productName;
  final double quantity;
  final double unitPrice;
  final double totalPrice;

  OrderItem({
    required this.id,
    this.productId,
    required this.productName,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    // Handle quantity as string or number
    double qty;
    if (json['quantity'] is String) {
      qty = double.parse(json['quantity'] as String);
    } else {
      qty = (json['quantity'] as num).toDouble();
    }
    
    return OrderItem(
      id: json['id'] as int,
      productId: json['productId'] as String?,
      productName: json['productName'] as String,
      quantity: qty,
      unitPrice: (json['unitPrice'] as num).toDouble(),
      totalPrice: (json['totalPrice'] as num).toDouble(),
    );
  }
  
  String get quantityDisplay {
    if (quantity < 1) {
      return '${(quantity * 1000).toInt()}g';
    } else if (quantity == quantity.toInt()) {
      return '${quantity.toInt()}';
    } else {
      return '${quantity}kg';
    }
  }
}

class OrderListResponse {
  final List<Order> orders;
  final int total;
  final int page;
  final int pageSize;

  OrderListResponse({
    required this.orders,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory OrderListResponse.fromJson(Map<String, dynamic> json) {
    return OrderListResponse(
      orders: (json['orders'] as List<dynamic>)
          .map((e) => Order.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['pageSize'] as int,
    );
  }
}

class OrderStats {
  final int totalOrders;
  final int todayOrders;
  final int pendingOrders;
  final double totalRevenue;

  OrderStats({
    required this.totalOrders,
    required this.todayOrders,
    required this.pendingOrders,
    required this.totalRevenue,
  });

  factory OrderStats.fromJson(Map<String, dynamic> json) {
    return OrderStats(
      totalOrders: json['totalOrders'] as int,
      todayOrders: json['todayOrders'] as int,
      pendingOrders: json['pendingOrders'] as int,
      totalRevenue: (json['totalRevenue'] as num).toDouble(),
    );
  }
}
